agent/checkpoint.go

```diff
@@ -0,0 +1,33 @@
+package agent
+
+import (
+	"context"
+)
+
+// checkpoint allows a goroutine to communicate when it is OK to proceed beyond some async condition
+// to other dependent goroutines.
+type checkpoint struct {
+	done chan struct{}
+	err  error
+}
+
+// complete the checkpoint.  Pass nil to indicate the checkpoint was ok.
+func (c *checkpoint) complete(err error) {
+	c.err = err
+	close(c.done)
+}
+
+func (c *checkpoint) waitCtx(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-c.done:
+		return c.err
```

Comment:
Should one of these have higher priority than the other?

Reply:
I don't think there is a need, and go doesn't make strong guarantees about how selects get ordered when multiple channels become readable close to the same time.

Reply:
Alright, just thought I'd mention in case it would matter. Prioritization is quite simple, albeit a bit verbose, essentially re-testing like this:

```go
	select {
	case <-ctx.Done():
		select {
		case <-c.done:
			return c.err
		default:
			return ctx.Err()
		}
	case <-c.done:
		return c.err
	}
```